File: app/src/main/java/eu/rtsketo/sakurastats/main/Console.java
package eu.rtsketo.sakurastats.main;

import android.app.Activity;

import com.qwerjk.better_text.MagicTextView;

import static eu.rtsketo.sakurastats.control.ViewDecor.decorate;
import static eu.rtsketo.sakurastats.hashmaps.SDPMap.sdp2px;

public class Console {
    private Activity activity;
    private MagicTextView mtv;
    private static Console instance;
    private StringBuilder log = new StringBuilder();

    private Console(MagicTextView mtv) {
        this.mtv = mtv;
    }

    private synchronized void appendText(String text, boolean newLine) {
        String[] lines  = log.toString().split("\n");
        if (lines.length > 20)
            log.replace(0, lines[0].length()+1, "");

        log.append(text);
        if (newLine) log.append("\n");
        activity.runOnUiThread(()->
                decorate(mtv, log.toString(), sdp2px(8)));
    }
    public static void logln(String text) {
        if (instance != null)
            instance.appendText(text, true);
    }

    public static void log(String text) {
        if (instance != null)
            instance.appendText(
                    text + " ", false);
    }

    public static void create(Activity activity, MagicTextView mtv) {
        instance = new Console(mtv);
        instance.activity = activity;
    }

    public static String convertRole(String role) {
        if (role == null) return "";
        switch (role) {
            case "coLeader": role = "Co-Leader "; break;
            case "leader": role = "Leader     "; break;
            case "elder": role = "Elder        "; break;
            default: role = "Member    ";
        }
        return role;
    }
}

File: app/src/main/java/eu/rtsketo/sakurastats/main/Console.kt
package eu.rtsketo.sakurastats.main

import android.app.Activity
import com.qwerjk.better_text.MagicTextView
import eu.rtsketo.sakurastats.control.ViewDecor.decorate
import eu.rtsketo.sakurastats.hashmaps.SDPMap

class Console private constructor(private val mtv: MagicTextView) {
    private var activity: Activity? = null
    private val log = StringBuilder()

    @Synchronized
    private fun appendText(text: String, newLine: Boolean) {
        val lines = log.toString().split("\n").toTypedArray()
        if (lines.size > 20) log.replace(0, lines[0].length + 1, "")
        log.append(text)
        if (newLine) log.append("\n")
        activity?.runOnUiThread {
            decorate(mtv, log.toString(),
                    SDPMap.sdp2px(8).toFloat()) }
    }

    companion object {
        private var instance: Console? = null
        fun logln(text: String) {
            instance?.appendText(text, true)
        }

        fun log(text: String) {
            instance?.appendText("$text ", false)
        }

        fun create(activity: Activity, mtv: MagicTextView) {
            instance = Console(mtv)
            instance!!.activity = activity
        }

        fun convertRole(role: String): String {
            return when (role) {
                "coLeader" -> "Co-Leader "
                "leader" -> "Leader     "
                "elder" -> "Elder        "
                else -> "Member    "
            }
        }
    }

}